### ISSUE TITLE
Leak FP when using a lambda capture

### ISSUE THREAD

**Issue by steakhal** [llvm/llvm-project#219810]
https://godbolt.org/z/W3Y847sMT
```c++
struct Ptr {
    int* p;
    explicit Ptr(int* p) : p(p) {}
    Ptr(Ptr&& o) : p(o.p) { o.p = nullptr; }
    ~Ptr() { delete p; }
};
void top() {
    auto l = [q = Ptr(new int)] {}; // FP: Potential leak of memory pointed to by field 'p' [cplusplus.NewDeleteLeaks]
    l();
}
```

A similar issue was in the past #60896 for unique_ptr leaks and those were suppressed in #152751 but that only suppresses for standard smart-pointers, thus a custom uptr like class like `Ptr` here still warns.

**Comment by Hari-Shankar-Karthik** [comment]
Hi! I'm working on this issue, would like to be assigned to it. Thanks!

**Comment by Hari-Shankar-Karthik** [comment]
By extension, there is also a False Negative when it comes to double-free:
```cpp
struct Ptr {
    int* p;
    explicit Ptr(int* p) : p(p) {}
    Ptr(Ptr&& o) : p(o.p) { o.p = nullptr; }
    ~Ptr() { delete p; }
};
void top() {
    int *p = new int;
    auto l = [q = Ptr(p)] {};
    l();
    delete p;
} // FN: Expect a double-free warning when l goes out of scope
```

Therefore, unlike the similar [smart pointer issue](https://github.com/llvm/llvm-project/issues/60896), the solution for this will probably require fixing the analyzer's machinery to properly model ownership transfer in such cases.

**Comment by steakhal** [comment]
If I had to guess the root cause of this is that:
when modeling the initializer list expression and the assigment in the ctor body, it confuses something and it decides to escape the member, thus forget that it held a heap pointer.


If my guess is correct, then fixing this is going to be difficult. This has came up a few times and had a coule attempts for properly gixing it.
